scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,395 @@
+    # larger than the largest element of the haystack, we
+    # explicitly add an element not in needles at the end.
+    haystack = np.concatenate([haystack, [-1]])
+    indexed = haystack[indices]
+    diff = indexed != needles
+    # At this point, diff contains the indices of the (j, i)
+    # to be added. We make use of the COO initializer for CSR
+    # matrices which ensures that 0s are added explicitly.
+    indices_to_add = np.where(diff)[0]
+    data_to_add = np.zeros((indices_to_add.shape[0],), dtype=np.int32)
+    acoo.row = np.concatenate([acoo.row, cols[indices_to_add]])
+    acoo.col = np.concatenate([acoo.col, rows[indices_to_add]])
+    acoo.data = np.concatenate([acoo.data, data_to_add])
+    # Repeat process to get reverse pointers
+    return acoo.tocsr()
+
+
+def _make_edge_pointers(a):
+    """Create for each edge pointers to its reverse and its tail."""
+    # Just as above, one way to match (i, j) with (j, i) is to match
+    # i*n + j with j*n + i, which we can do by ensuring that we
+    # have a sorted collection of {i*n + j} and using binary search
+    # to find j*n + i here.
+    n = a.shape[0]
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    rev_edge_ptr = np.searchsorted(haystack, needles).astype(np.int32)
+    return rev_edge_ptr, rows
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef int[:] _edmonds_karp(
+        int[:] edge_ptr,
+        int[:] tails,
+        int[:] heads,
+        int[:] capacities,
+        int[:] rev_edge_ptr,
+        int source,
+        int sink):
+    """Solves the maximum flow problem using the Edmonds--Karp algorithm.
+
+    This assumes that for every edge in the graph, the edge in the opposite
+    direction is also in the graph (possibly with capacity 0).
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length |V| + 1
+        For a given vertex v, the edges whose tail is v are those between
+        edge_ptr[v] and edge_ptr[v + 1] - 1.
+    tails : memoryview of length |E|
+        For a given edge e, tails[e] is the tail vertex of e.
+    heads : memoryview of length |E|
+        For a given edge e, tails[e] is the head vertex of e.
+    capacities : memoryview of length |E|
+        For a given edge e, capacities[e] is the capacity of e.
+    rev_edge_ptr : memoryview of length |E|
+        For a given edge e, rev_edge_ptr[e] is the edge obtained by
+        reversing e. In particular, rev_edge_ptr[rev_edge_ptr[e]] == e.
+    source : int
+        The source vertex.
+    sink : int
+        The sink vertex.
+
+    Returns
+    -------
+    flow : memoryview of length |E|
+        The residual graph with respect to a maximum flow.
+
+    """
+    cdef int n_verts = edge_ptr.shape[0] - 1
+    cdef int n_edges = capacities.shape[0]
+
+    # Our result array will keep track of the flow along each edge
+    cdef int[:] flow = np.zeros(n_edges, dtype=np.int32)
```

Comment:
The rest of the `csgraph` module uses `ITYPE` instead of `np.int32`, it would be good to be consistent. Doesn't change how the code works right now, but helps with maintenance. For context: https://github.com/scipy/scipy/pull/243#issuecomment-6233652

Reply:
Makes sense. I updated this in the Cython functions, the wrapping pure-Python functions, but not in the example using `np.int`.

Reply:
Looks good module my comment on the hardcoded max value